_includes/tutorials/aggregating-average/kstreams/code/src/main/java/io/confluent/developer/RunningAverage.java

```diff
@@ -0,0 +1,211 @@
+import static org.apache.kafka.streams.StreamsConfig.BOOTSTRAP_SERVERS_CONFIG;
+import static org.apache.kafka.streams.StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG;
+import static org.apache.kafka.streams.StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG;
+import static org.apache.kafka.streams.StreamsConfig.REPLICATION_FACTOR_CONFIG;
+import static org.apache.kafka.streams.kstream.Grouped.with;
+
+public class RunningAverage {
+
+  //region buildStreamsProperties
+  protected Properties buildStreamsProperties(Properties envProps) {
+    Properties config = new Properties();
+    config.putAll(envProps);
+
+    config.put(APPLICATION_ID_CONFIG, envProps.getProperty("application.id"));
+    config.put(BOOTSTRAP_SERVERS_CONFIG, envProps.getProperty("bootstrap.servers"));
+    config.put(DEFAULT_KEY_SERDE_CLASS_CONFIG, Long().getClass());
+    config.put(DEFAULT_VALUE_SERDE_CLASS_CONFIG, Double().getClass());
+    config.put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, envProps.getProperty("schema.registry.url"));
+
+    config.put(REPLICATION_FACTOR_CONFIG, envProps.getProperty("default.topic.replication.factor"));
+    config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, envProps.getProperty("offset.reset.policy"));
+
+    config.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+
+    return config;
+  }
+  //endregion
+
+  //region createTopics
+
+  /**
+   * Create topics using AdminClient API
+   */
+  private void createTopics(Properties envProps) {
+    Map<String, Object> config = new HashMap<>();
+
+    config.put("bootstrap.servers", envProps.getProperty("bootstrap.servers"));
+    AdminClient client = AdminClient.create(config);
+
+    List<NewTopic> topics = new ArrayList<>();
+
+    topics.add(new NewTopic(
+        envProps.getProperty("input.ratings.topic.name"),
+        parseInt(envProps.getProperty("input.ratings.topic.partitions")),
+        parseShort(envProps.getProperty("input.ratings.topic.replication.factor"))));
+
+    topics.add(new NewTopic(
+        envProps.getProperty("output.rating-averages.topic.name"),
+        parseInt(envProps.getProperty("output.rating-averages.topic.partitions")),
+        parseShort(envProps.getProperty("output.rating-averages.topic.replication.factor"))));
+
+    client.createTopics(topics);
+    client.close();
+
+  }
+  //endregion
+
+  private void run() {
+
+    Properties envProps = this.loadEnvProperties();
+    Properties streamProps = this.buildStreamsProperties(envProps);
+    Topology topology = this.buildTopology(new StreamsBuilder(), envProps);
+
+    this.createTopics(envProps);
+
+    final KafkaStreams streams = new KafkaStreams(topology, streamProps);
+    final CountDownLatch latch = new CountDownLatch(1);
+
+    // Attach shutdown handler to catch Control-C.
+    Runtime.getRuntime().addShutdownHook(new Thread("streams-shutdown-hook") {
+      @Override
+      public void run() {
+        streams.close();
```

Comment:
running the test harness locally the streams app wouldn't shut down changing `streams.close()` to `streams.close(Duration.ofSeconds(5));` fixes that.  Otherwise, the main thread will block forever, waiting for streams to stop.

Reply:
this one nice @bbejeck 
Thanks

Reply:
done